File: src/shared/types/update.ts
export type UpdateStatus =
  | "idle"
  | "checking"
  | "up_to_date"
  | "available"
  | "downloading"
  | "downloaded"
  | "installing"
  | "error";

export type UpdateErrorStage =
  | "check"
  | "download"
  | "install";

export interface UpdateSnapshot {
  currentVersion: string;
  status: UpdateStatus;
  latestVersion: string | null;
  releaseNotes: string | null;
  releaseDate: string | null;
  errorMessage: string | null;
  errorStage: UpdateErrorStage | null;
  downloadedBytes: number | null;
  totalBytes: number | null;
  releasePageUrl: string | null;
  assetDownloadUrl: string | null;
}

export function createIdleUpdateSnapshot(currentVersion = "0.0.0"): UpdateSnapshot {
  return {
    currentVersion,
    status: "idle",
    latestVersion: null,
    releaseNotes: null,
    releaseDate: null,
    errorMessage: null,
    errorStage: null,
    downloadedBytes: null,
    totalBytes: null,
    releasePageUrl: null,
    assetDownloadUrl: null,
  };
}

File: src-tauri/src/domain/update.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum UpdateStatus {
    Idle,
    Checking,
    UpToDate,
    Available,
    Downloading,
    Downloaded,
    Installing,
    Error,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateSnapshot {
    pub current_version: String,
    pub status: UpdateStatus,
    pub latest_version: Option<String>,
    pub release_notes: Option<String>,
    pub release_date: Option<String>,
    pub error_message: Option<String>,
}

impl UpdateSnapshot {
    pub fn idle(current_version: String) -> Self {
        Self {
            current_version,
            status: UpdateStatus::Idle,
            latest_version: None,
            release_notes: None,
            release_date: None,
            error_message: None,
        }
    }
}
